Add new deps in 2025

toyboot4e opened the pull request:
Here we go~

Edit: Sorry I missed package descriptions.

- `ac-library-hs` ([GitHub](https://github.com/toyboot4e/ac-library-hs), [Hackage](https://hackage.haskell.org/package/ac-library-hs))
- `ilist` ([GitHub](https://github.com/brandonhamilton/ilist), [Hackage](https://hackage.haskell.org/package/ilist))
- `monad-memo` ([GitHub](https://github.com/EduardSergeev/monad-memo), [Hackage](https://hackage.haskell.org/package/monad-memo))
- `vector-split` ([GitHub](https://github.com/fhaust/vector-split), [Hackage](https://hackage.haskell.org/package/vector-split))
- `wide-word` ([GitHub](https://github.com/erikd/wide-word), [Hackage](https://hackage.haskell.org/package/wide-word))
- `witherable` ([GitHub](https://github.com/fumieval/witherable), [Hackage](https://hackage.haskell.org/package/witherable))

`hmatrix` requires additional `apt` dependencies, as described in [`INSTALL.md`](https://github.com/haskell-numerics/hmatrix/blob/master/INSTALL.md).

- `hmatrix` ([GitHub](https://github.com/haskell-numerics/hmatrix), [Hackage](https://hackage.haskell.org/package/hmatrix))
- `hmatrix-glpk` ([Hackage](https://hackage.haskell.org/package/hmatrix-glpk))
- `hmatrix-gsl` ([Hackage](https://hackage.haskell.org/package/hmatrix-gsl))
- `hmatrix-special` ([Hackage](https://hackage.haskell.org/package/hmatrix-special))

There are [some other `hmatrix-*` libraries as searched in Hackage](https://hackage.haskell.org/packages/search?terms=hmatrix). I didn't look into them (yet).

gksato commented:
Thanks, I'm reading this!

gksato commented on Makefile:
```diff
@@ -11,7 +11,8 @@ cabalver = 3.12.1.0
 ghcup_apt_dependency = build-essential curl libffi-dev libffi8ubuntu1 \
-	libgmp-dev libgmp10 libncurses-dev libncurses6 libtinfo6 llvm-15
+	libgmp-dev libgmp10 libncurses-dev libncurses6 libtinfo6 llvm-15 \
+	libgsl0-dev liblapack-dev libatlas-base-dev libglpk-dev
```
This line's bugging me for maintainability and rationale. I know that apt-installation should be done in one place for readability, but ghcup Dockerfile and serverproto Dockerfile are separated to offer toolgen a vanilla workspace (a weak claim that apt-installation should be moved to serverproto).

For maintainability, whether or not the apt-installation for `hmatrix` is moved to serverproto Dockerfile after the GHCup installation, there should be another variable. We need to make clear that the reason for the packages is not GHCup but hmatrix.

gksato replied:
The reason why I'm saying my first suggestion is weak is because we need to place apt-installation after the installation of GHCup in installation script to make Docker image exactly immitate the installation. Docker images are to immitate the installation, but this is making things other way round; We're bending the installation to align it with the immitation.

toyboot4e replied:
I guess it's OK to run `apt-get` twice on the AtCoder server, as long as the installation script is kept clear. I'll try moving the `hmatrix` installation from ghcup `Dockerfile` to the serverproto `Dockerfile` with my zero Docker skill.

toyboot4e commented:
So I moved `hmatrix` installation to serverproto `Dockerfile` and changed `install.sh.template` to match the instructions in the Dockerfiles. Is this change what you meant? Thank you.

gksato commented:
I guess so, but I'll make sure after I recover from dizziness, tomorrow morning that is.

toyboot4e commented:
I rushed too much, please take your time.

gksato commented:
Everything looks neatly done, I'm running everything from the Makefile!

gksato commented:
Nice, merging!
